content/blog/handling-react-remote-modules-with-babel.md

```diff
@@ -0,0 +1,267 @@
+
+It seemed to us like this would be a `babel` type of problem. We wanted to interpret the code and transform it. Abstract Syntax Trees, I'm looking at you :eyes:
+
+As the project we're currently doing this was configured using webpack, our starting point was the [babel-loader](https://webpack.js.org/loaders/babel-loader/) documentation page. We understood that we needed to create a custom loader, and load in from webpack's configuration. Because the project uses TypeScript, we needed to register this babel loader to run after `ts-loader`.
+
+This is how it looked on webpack configuration.
+
+```js
+
+const webpackConfig = {
+  entry: [ /* entry files */],
+  resolve: { /* resolve config */ },
+  plugins: [/* plugins import */ ],
+  module: {
+    rules: [
+      {
+        test: /\.(tsx?)?$/,
+        include: [ path.resolve('src'), ],
+        use: [
+          {
+            loader: require.resolve('./loader.js'),
+          },
+          {
+            loader: 'ts-loader',
+          },
+        ],
+        exclude: /node_modules/,
+      },
+  },
+  // cut for brevity
+};
+```
+
+We used `require.resolve` to load our `loader.js` file that was on the same folder as this configuration.
+
+From there, our adventure writing a babel plugin to navigate and modify the AST started. [AST Explorer]() was of great use in this step, we used it to develop and debug the script we were writing.
+
+Just doing a small recap, we wanted to do the following transformation:
+
+This:
+
+```js
+import Button from "shared/components/Button";
+```
+
+To this
+
+```js
+const Button = React.lazy(() => import("shared/components/Button"));
+```
+
+## The plugin
+
+By adding the initial code on [AST Explorer], we could already see the AST of it, we created a custom babel loader, opened [babel-handbook], and here we go:
+
+This was our initial code.
+
+```js
+
+module.exports = function(babel) {
+  const {types: t} = babel;
+
+  function moduleFederationReactImportAliases(_, ...options) {
+    return {
+      visitor: {
+        }
+      },
+    };
+  }
+
+  return {
+    config(cfg) {
+      return {
+        ...cfg.options,
+        plugins: [
+          ...(cfg.options.plugins || []),
+          moduleFederationReactImportAliases
+        ],
+      };
+    },
+
+    result(result) {
+      return {
+        ...result,
+        code: result.code
+      };
+    },
+  };
+};
+```
+
+We knew that babel-loaders needed to use the `babel-config` `custom` function with a callback, and that they would need to register themselves as a plugin in the global configuration.
+
+As we skimmed through babel-handbook, we also noted that we'd most likely use the visitor pattern, starting by _visiting_ the `ImportDeclarations`, since our initial code was one of them. We decided that we wanted every import starting with `shared` to be a target for this transformation, and thus we added the code to do that.
```

Comment:
And we don't want `every` shared component to do this, only the ones we register in the options